File: backend/internal/service/provision.go
package service

import "time"

type ProvisionPlan struct {
	ID            int64     `json:"id"`
	Code          string    `json:"code"`
	Name          string    `json:"name"`
	GroupID       int64     `json:"group_id"`
	Balance       float64   `json:"balance"`
	Quota         float64   `json:"quota"`
	ExpiresInDays *int      `json:"expires_in_days,omitempty"`
	RateLimit5h   float64   `json:"rate_limit_5h"`
	RateLimit1d   float64   `json:"rate_limit_1d"`
	RateLimit7d   float64   `json:"rate_limit_7d"`
	Concurrency   int       `json:"concurrency"`
	RPMLimit      int       `json:"rpm_limit"`
	Enabled       bool      `json:"enabled"`
	Group         *Group    `json:"group,omitempty"`
	CreatedAt     time.Time `json:"created_at"`
	UpdatedAt     time.Time `json:"updated_at"`
}

type ProvisionPlanInput struct {
	Code          string  `json:"code"`
	Name          string  `json:"name"`
	GroupID       int64   `json:"group_id"`
	Balance       float64 `json:"balance"`
	Quota         float64 `json:"quota"`
	ExpiresInDays *int    `json:"expires_in_days"`
	RateLimit5h   float64 `json:"rate_limit_5h"`
	RateLimit1d   float64 `json:"rate_limit_1d"`
	RateLimit7d   float64 `json:"rate_limit_7d"`
	Concurrency   int     `json:"concurrency"`
	RPMLimit      int     `json:"rpm_limit"`
	Enabled       bool    `json:"enabled"`
}

type ProvisionOrder struct {
	ID            int64             `json:"id"`
	OrderID       string            `json:"order_id"`
	PlanID        *int64            `json:"plan_id,omitempty"`
	PlanCode      string            `json:"plan_code"`
	PlanSnapshot  ProvisionSnapshot `json:"plan_snapshot"`
	UserID        *int64            `json:"user_id,omitempty"`
	APIKeyID      *int64            `json:"api_key_id,omitempty"`
	Status        string            `json:"status"`
	CustomerLabel string            `json:"customer_label"`
	CreatedAt     time.Time         `json:"created_at"`
	UpdatedAt     time.Time         `json:"updated_at"`
}

type ProvisionSnapshot struct {
	PlanID         int64   `json:"plan_id"`
	PlanCode       string  `json:"plan_code"`
	PlanName       string  `json:"plan_name"`
	GroupID        int64   `json:"group_id"`
	GroupName      string  `json:"group_name"`
	Platform       string  `json:"platform"`
	RateMultiplier float64 `json:"rate_multiplier"`
	Balance        float64 `json:"balance"`
	Quota          float64 `json:"quota"`
	ExpiresInDays  *int    `json:"expires_in_days,omitempty"`
	RateLimit5h    float64 `json:"rate_limit_5h"`
	RateLimit1d    float64 `json:"rate_limit_1d"`
	RateLimit7d    float64 `json:"rate_limit_7d"`
	Concurrency    int     `json:"concurrency"`
	RPMLimit       int     `json:"rpm_limit"`
}

type ProvisionResult struct {
	OrderID        string  `json:"order_id"`
	APIKey         string  `json:"api_key"`
	KeyID          int64   `json:"key_id"`
	UserID         int64   `json:"user_id"`
	PlanCode       string  `json:"plan_code"`
	GroupID        int64   `json:"group_id"`
	Balance        float64 `json:"balance"`
	Quota          float64 `json:"quota"`
	RateMultiplier float64 `json:"rate_multiplier"`
}

type ProvisionAPIKeyInput struct {
	OrderID       string `json:"order_id"`
	PlanCode      string `json:"plan_code"`
	CustomerLabel string `json:"customer_label"`
}

File: frontend/src/api/admin/provision.ts
import { apiClient } from '../client'
import type { AdminGroup } from '@/types'

function randomIdempotencyKey(scope: string): string {
  const random =
    typeof crypto !== 'undefined' && typeof crypto.randomUUID === 'function'
      ? crypto.randomUUID()
      : `${Date.now()}-${Math.random().toString(36).slice(2)}`
  return `${scope}:${random}`
}

function provisionIdempotencyKey(payload: ProvisionAPIKeyRequest): string {
  return `provision-api-key:${payload.order_id.trim()}:${payload.plan_code.trim().toLowerCase()}`
}

function idempotencyHeaders(key: string) {
  return {
    headers: {
      'Idempotency-Key': key
    }
  }
}

export interface ProvisionPlan {
  id: number
  code: string
  name: string
  group_id: number
  balance: number
  quota: number
  expires_in_days?: number | null
  rate_limit_5h: number
  rate_limit_1d: number
  rate_limit_7d: number
  concurrency: number
  rpm_limit: number
  enabled: boolean
  group?: AdminGroup
  created_at: string
  updated_at: string
}

export interface ProvisionPlanRequest {
  code: string
  name: string
  group_id: number
  balance: number
  quota: number
  expires_in_days?: number | null
  rate_limit_5h: number
  rate_limit_1d: number
  rate_limit_7d: number
  concurrency: number
  rpm_limit: number
  enabled: boolean
}

export interface ProvisionAPIKeyRequest {
  order_id: string
  plan_code: string
  customer_label?: string
}

export interface ProvisionResult {
  order_id: string
  api_key: string
  key_id: number
  user_id: number
  plan_code: string
  group_id: number
  balance: number
  quota: number
  rate_multiplier: number
}

export async function listPlans(): Promise<ProvisionPlan[]> {
  const { data } = await apiClient.get<ProvisionPlan[]>('/admin/provision/plans')
  return data || []
}

export async function createPlan(payload: ProvisionPlanRequest): Promise<ProvisionPlan> {
  const { data } = await apiClient.post<ProvisionPlan>(
    '/admin/provision/plans',
    payload,
    idempotencyHeaders(randomIdempotencyKey('provision-plan-create'))
  )
  return data
}

export async function updatePlan(id: number, payload: ProvisionPlanRequest): Promise<ProvisionPlan> {
  const { data } = await apiClient.put<ProvisionPlan>(
    `/admin/provision/plans/${id}`,
    payload,
    idempotencyHeaders(randomIdempotencyKey(`provision-plan-update:${id}`))
  )
  return data
}

export async function deletePlan(id: number): Promise<{ message: string }> {
  const { data } = await apiClient.delete<{ message: string }>(`/admin/provision/plans/${id}`)
  return data
}

export async function provisionAPIKey(payload: ProvisionAPIKeyRequest): Promise<ProvisionResult> {
  const { data } = await apiClient.post<ProvisionResult>(
    '/admin/provision/api-keys',
    payload,
    idempotencyHeaders(provisionIdempotencyKey(payload))
  )
  return data
}

export default {
  listPlans,
  createPlan,
  updatePlan,
  deletePlan,
  provisionAPIKey
}
